File: app/src/main/java/net/chmielowski/fiszki/MainActivity.java
package net.chmielowski.fiszki;

import android.content.Context;
import android.media.MediaPlayer;
import android.os.Bundle;
import android.os.Vibrator;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.annimon.stream.Optional;
import com.annimon.stream.Stream;

import java.io.IOException;
import java.util.List;
import java.util.Random;

public class MainActivity extends AppCompatActivity {

    public static final String LANGUAGE = "language";
    public static final String NUMBER_OF_WORDS = "number_of_words";
    private static final String LESSON_ID = "LESSON_ID";
    private final int NUMBER_OF_REPETITIONS = 3;
    private Random random = new Random();
    private Word word;
    private RealmDelegate realmDelegate = new RealmDelegate();
    private Lesson lesson;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        realmDelegate.onCreate();
        Optional.ofNullable(savedInstanceState)
                .map(state -> state.getSerializable(LESSON_ID))
                .executeIfPresent(lessonId -> {
                    // get lesson from DB and save to MainActivity.this.lesson
                })
                .executeIfAbsent(() -> {
                    lesson = DictionaryUtils.getLesson(
                            getLanguage(), getNumberOfWords(), realmDelegate.getRealm());
//                    words = Collections.unmodifiableList(lesson.getGroups());
                });

        final Vibrator v = (Vibrator) getSystemService(Context.VIBRATOR_SERVICE);
        findViewById(R.id.show).setOnClickListener(view -> {
            vibrate(v);
            showAnswer();
        });
        findViewById(R.id.pass).setOnClickListener(view -> {
            vibrate(v);
            onPass();
        });
        findViewById(R.id.fail).setOnClickListener(view -> {
            vibrate(v);
            onFail();
        });
        ((ProgressBar) findViewById(R.id.progress)).setMax(lesson.numberOfAllWords() * NUMBER_OF_REPETITIONS);
        nextWord();
        findViewById(R.id.play).setOnClickListener(view -> {
            vibrate(v);
            new Thread(() -> {
                try {
                    final MediaPlayer player = new MediaPlayer();
                    player.setDataSource(String.format(
                            "https://translate.google.com/translate_tts?ie=UTF-8&q=%s&tl=%s&client=tw-ob",
                            word.foreign,
                            getLanguage().shortcut
                    ));
                    player.prepareAsync();
                    player.setOnPreparedListener(MediaPlayer::start);
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            }).start();
        });
    }

    private int getNumberOfWords() {
        return getIntent().getIntExtra(NUMBER_OF_WORDS, NUMBER_OF_REPETITIONS);
    }

    private DictionaryUtils.Lang getLanguage() {
        return (DictionaryUtils.Lang) getIntent().getSerializableExtra(LANGUAGE);
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        outState.putString(LESSON_ID, lesson.id);
        super.onSaveInstanceState(outState);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        realmDelegate.onDestroy();
    }

    private void vibrate(Vibrator v) {v.vibrate(20);}

    private void onFail() {
        nextWord();
    }

    private void onPass() {
        realmDelegate.getRealm()
                     .executeTransaction(db -> lesson.incrementScoreOf(word)
                     );
        nextWord();
    }

    private void showAnswer() {
        findViewById(R.id.foreign).setVisibility(View.VISIBLE);
        enable(R.id.pass);
        enable(R.id.fail);
        enable(R.id.play);
        disable(R.id.show);
    }

    private Integer score() {
        return lesson.score();
    }

    private void nextWord() {
        updateProgressBar();
        chooseNext();
        ((TextView) findViewById(R.id.english)).setText(word.english);
        ((TextView) findViewById(R.id.foreign)).setText(word.foreign);
        findViewById(R.id.foreign).setVisibility(View.INVISIBLE);
        disable(R.id.pass);
        disable(R.id.fail);
        disable(R.id.play);
        enable(R.id.show);
    }

    private void chooseNext() {
        final List<Word> unknown = Stream.of(lesson.scores)
                                         .filter(score -> score.getScore() < NUMBER_OF_REPETITIONS)
                                         .map(score -> score.word)
                                         .toList();
        final boolean finished = unknown.size() == 0;
        if (finished) {
            realmDelegate.getRealm()
                         .executeTransaction(db -> Stream.of(lesson.groups)
                                                         .forEach(w -> w.score++));
            finish();
            return;
        }
        word = unknown.get(random.nextInt(unknown.size()));
    }

    private void updateProgressBar() {
        ((ProgressBar) findViewById(R.id.progress)).setProgress(score());
    }

    private void enable(int view) {
        findViewById(view).setEnabled(true);
        findViewById(view).setAlpha(1f);
    }

    private void disable(int view) {
        findViewById(view).setAlpha(.2f);
        findViewById(view).setEnabled(false);
    }

}

File: app/src/main/java/net/chmielowski/fiszki/MainActivity.kt
package net.chmielowski.fiszki

import android.content.Context
import android.media.MediaPlayer
import android.os.Bundle
import android.os.Vibrator
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.widget.ProgressBar
import android.widget.TextView
import java.util.*


class MainActivity : AppCompatActivity() {
    private val NUMBER_OF_REPETITIONS = 2
    private val myView = MyView(this)
    private val random = Random()
    private lateinit var word: Word
    private val realmDelegate = RealmDelegate()
    private lateinit var lesson: Lesson

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        RoomService().saveData(applicationContext)
        setContentView(R.layout.activity_main)
        realmDelegate.onCreate()
        lesson = savedInstanceState
                ?.getString(LESSON_ID)
                ?.let {
                    realmDelegate.realm
                            .where(Lesson::class.java)
                            .equalTo("id", it)
                            .findFirst()
                }
                ?: DictionaryUtils.getLesson(
                language,
                numberOfWords,
                realmDelegate.realm)
        val v = getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
        findViewById<View>(R.id.show).setOnClickListener {
            vibrate(v)
            myView.showAnswer()
        }
        findViewById<View>(R.id.pass).setOnClickListener {
            vibrate(v)
            onPass()
        }
        findViewById<View>(R.id.fail).setOnClickListener {
            vibrate(v)
            onFail()
        }
        findViewById<ProgressBar>(R.id.progress).max = lesson.numberOfAllWords() * NUMBER_OF_REPETITIONS
        nextWord()
        findViewById<View>(R.id.play).setOnClickListener {
            vibrate(v)
            play(word.foreign, language.shortcut)
        }
    }


    private val numberOfWords: Int
        get() = intent.getIntExtra(NUMBER_OF_WORDS, NUMBER_OF_REPETITIONS)

    private val language: DictionaryUtils.Lang
        get() = intent.getSerializableExtra(LANGUAGE) as DictionaryUtils.Lang

    override fun onSaveInstanceState(outState: Bundle?) {
        outState!!.putString(LESSON_ID, lesson.id)
        super.onSaveInstanceState(outState)
    }

    override fun onDestroy() {
        super.onDestroy()
        realmDelegate.onDestroy()
    }

    private fun vibrate(v: Vibrator) {
        @Suppress("DEPRECATION")
        v.vibrate(20)
    }

    private fun onFail() {
        nextWord()
    }

    private fun onPass() {
        realmDelegate.realm
                .executeTransaction { lesson.incrementScoreOf(word) }
        nextWord()
    }

    private fun score() = lesson.score()

    private fun nextWord() {
        chooseNext(andThen = Runnable { myView.refreshView() })
    }

    private fun chooseNext(andThen: Runnable) {
        val unknown = lesson.scores
                .filter { it.score < NUMBER_OF_REPETITIONS }
                .map { it.word }
                .toList()
        val finished = unknown.isEmpty()
        if (finished) {
            realmDelegate.realm
                    .executeTransaction { db ->
                        lesson.groups.forEach { it.score++ }
                        db.delete(Lesson::class.java)
                        db.delete(WordScore::class.java)
                    }
            finish()
            return
        }
        word = unknown[random.nextInt(unknown.size)]
        andThen.run()
    }

    private class MyView internal constructor(private val activity: MainActivity) {

        private fun disable(view: Int) {
            this.activity.findViewById<View>(view).alpha = .2f
            this.activity.findViewById<View>(view).isEnabled = false
        }

        private fun enable(view: Int) {
            this.activity.findViewById<View>(view).isEnabled = true
            this.activity.findViewById<View>(view).alpha = 1f
        }

        private fun updateProgressBar() {
            (this.activity.findViewById<View>(R.id.progress) as ProgressBar).progress = this.activity.score()
        }

        internal fun refreshView() {
            updateProgressBar()
            this.activity.findViewById<TextView>(R.id.english).text = this.activity.word.english
            this.activity.findViewById<TextView>(R.id.foreign).text = this.activity.word.foreign
            this.activity.findViewById<View>(R.id.foreign).visibility = View.INVISIBLE
            disable(R.id.pass)
            disable(R.id.fail)
            disable(R.id.play)
            enable(R.id.show)
        }

        internal fun showAnswer() {
            this.activity.findViewById<View>(R.id.foreign).visibility = View.VISIBLE
            enable(R.id.pass)
            enable(R.id.fail)
            enable(R.id.play)
            disable(R.id.show)
        }
    }

    companion object {
        val LANGUAGE = "language"
        val NUMBER_OF_WORDS = "number_of_words"
        private val LESSON_ID = "LESSON_ID"
    }
}

fun play(word: String, language: String) {
    Thread {
        with(MediaPlayer()) {
            setDataSource(String.format(
                    "https://translate.google.com/translate_tts?ie=UTF-8&q=%s&tl=%s&client=tw-ob",
                    word,
                    language
            ))
            prepareAsync()
            setOnPreparedListener { it.start() }
        }
    }.start()
}
